keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,604 @@
+
+### Risks and Mitigations
+
+- If the job object's size reaches to limit of the etcd and
+the job controller can't store a correct value in `.status.completedIndexes`,
+we probably can not evaluate the SuccessPolicy correctly.
+
+- If we allow to set unlimited size of the value in `.spec.successPolicy.onSucceededPods.indexes`,
+we have a risk similar to [KEP-3850: Backoff Limits Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/tree/76dcd4f342cc0388feb085e685d4cc018ebe1dc9/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs#risks-and-mitigations).
+So, we limit the size of `indexes` to 64Ki.
+
+## Design Details
+
+### Job API
+
+We extend the Job API to set different policies by which a Job can be declared as succeeded.
+
+```golang
+type JobSpec struct {
+	...
+	// Specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// This policy can not be specified more than one.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +optional
+	SuccessPolicy *SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some Indexes.
+type SuccessPolicy struct {
+	// Represents the requirement on the succeeded pods.
+	// +optional
+	OnSucceededPods *SuccessPolicyOnSucceededPods
+}
+
+// SuccessPolicyOnSucceededPods describes a Job can be succeeded based on succeeded pods.
+type SuccessPolicyOnSucceededPods struct {
+	// Specifies a set of required indexes when .spec.completionMode =
+	// "Indexed" in a text format.
+	// The job is declared as succeeded if a set of indexes are succeeded. 
+	// The list of indexes must come within 0 to `.spec.completions` and
+	// must not contain duplicates. At least one element is required. 
+	// The indexes are represented as decimal integers
+	// separated by commas. The numbers are listed in increasing order. Three or
+	// more consecutive numbers are compressed and represented by the first and
+	// last element of the series, separated by a hyphen.
+	// For example, if the completed indexes are 1, 3, 4, 5 and 7, they are
+	// represented as "1,3-5,7".
+	// 
+	// +optional
+	Indexes string
+
+	// Specifies the required number of indexes when .spec.completionMode =
+	// "Indexed".
+	// Value can be an absolute number (ex: 5) or a percentage of total indexes
+	// when the job can be declared as succeeded (ex: 50%).
+	// The absolute number is calculated from the percentage by rounding up.
+	// 
+	// +optional
+	Count intstr.IntOrString
+
+	// Specifies the action to be taken on the not finished (succeeded or failed) pods 
+	// when the job achieved the requirements.
+	// Possible values are:
+	//
+	// - Continue: indicates that all pods wouldn't be terminated.
```

Comment:
As these pods continue, will PodFailurePolicy still apply and potentially restart?

Reply:
> As these pods continue, will PodFailurePolicy still apply and potentially restart?

Yes, it's almost correct. However, I think that we shouldn't allow applying the PodFailurePolicy with the `action=FailJob` after the job is marked as successful by the successPolicies.

Because after marking the job as successful by suceessPolicies, we only allow the remaining indexes to keep running until the indexes are completed or failed, and we don't want to expect that to change the results of a job that has finalized results (completed/succeeded or failed) as you described in https://github.com/kubernetes/enhancements/pull/4062/files#r1458047126.

> once marked successful, the job should never be marked Failed.

cc: @mimowo @kannon92

Reply:
This does seem like a bit of leakage to me. We are saying that our Job is successful and we want to continue running. So I think we need to spell out both `PodFailurePolicy` and `backoffLimit` behavior. So it seems that if the rule is matched with continue, we would short circuit any failure case and mark the overall job as successful? So then we are skipping both `backoffLimit` and `PodFailurePolicy`?

I almost wonder what is the point of continue then?

Reply:
That sounds good to me. So the suggestion is to add a condition for when "SuccessPolicy" is matched and the job continues to run. But if there is any new failure we can still mark the job as Failed even if SuccessPolicy matched (ie backOffLimit, PodFailurePolicy).

Reply:
> I'd argue that continue doesn't mean finish what you've started, and ignore failures. Continue is continue as it would normally be. You just introduce an earlier success mode, but the rest should continue running as before. Otherwise the API is hard to follow.

It sounds reasonable to me. I think that once we introduce `Succeeded` condition type separate from `Completed`,
the job-controller can mark jobs failed by backoffLimit or podFailureCondition after a job meets SuccessPolicy with `whenTargetAchieved=continue`.

Reply:
We should think about how we'll teach this, and what something like `kubectl describe` would show for a job that has succeeded (condition) and also failed (Job outcome).

There are various ways to tweak around this. For example:
- rename the condition to `SuccessCriteriaMet` (which doesn't so obviously conflict with Failed).
- also set a `Succeeded` condition, but only when we know that the Job outcome can't become Failed.
- prohibit using the FailJob action in a way that a Job could set the succeeded condition and also fail overall

Reply:
To me, the natural interpretation of `continue` that once the success criteria are met, we let the running pods to continue, but ignore failures (don't recreate them based on backoffLimit or pod failure policy).

IIUC this is the intention of the story 2: you run a simulation, and have some minimal requirements on the number of successful (say 1000) samples to get statistical significance. Once you reach the threshold, you may prefer to let the extra samples finish to have a stronger signal, but you don't want to spend much more time / resources on pod restarts.

Also, I think respecting pod failure policy action `Count`, but not `FailJob` could be confusing.

Having said that, maybe there is room for 3 options: `continue`, `terminate` and `continueWithRecreations`. Given this is late, maybe we can do only terminate in the first Alpha iteration, but add a graduation point for Beta to decide on the set of supported options?

Reply:
What can we do to get this KEP merged at least as provisional? Let's aim for that and iterate towards alpha.

Reply:
> Also, I think respecting pod failure policy action Count, but not FailJob could be confusing.

It sounds reasonable, IIRC, @kannon92 also has the same concern.

> Having said that, maybe there is room for 3 options: continue, terminate and continueWithRecreations. Given this is late, maybe we can do only terminate in the first Alpha iteration, but add a graduation point for Beta to decide on the set of supported options?

This is a great point. I agree with adding only `terminate` to `whenTargetAchived` in the alpha stage.
I think we need to have more discussions on `continuing` action. However, I don't think that we have enough time for discussions in the alpha stage.

Reply:
+1 on `continuing` for alpha and we discuss more `continuing`. I think `terminate` is going to be very useful by itself so I think the KEP can proceed with just that option.

Reply:
Done.

Reply:
I agree with everyone, that the exit criteria require more thought, as we see every one of us is looking at it from a different angle. So making sure we all agree on consistent naming and conditions is paramount to make this kep a working thing.